torch/_prims/__init__.py

```diff
@@ -0,0 +1,566 @@
+
+cosh = _make_elementwise_unary_op(torch.cosh)
+
+
+#
+# Elementwise binary operations
+#
+
+add = _make_elementwise_binary_op(torch.add)
+
+atan2 = _make_elementwise_binary_op(torch.atan2)
+
+div = _make_elementwise_binary_op(torch.true_divide)
+
+mul = _make_elementwise_binary_op(torch.mul)
+
+sub = _make_elementwise_binary_op(torch.sub)
+
+
+# View operations
+
+def _broadcast_in_dim_checks(
+  a: Tensor,
+  shape: Sequence[int],
+  broadcast_dimensions: Sequence[int]):
+    # Type checks
+    assert isinstance(a, Tensor)
+    assert isinstance(shape, Sequence)
+    assert isinstance(broadcast_dimensions, Sequence)
+
+    # every dimension must be accounted for
+    assert a.ndim == len(broadcast_dimensions)
+
+    # broadcast shape must have weakly more dimensions
+    assert len(shape) >= a.ndim
+
+    # broadcast_dimensions must be an ascending sequence
+    # (no relative reordering of dims) of integers and
+    # each dimension must be within the new shape
+    def _greater_than_reduce(acc, x):
+        assert isinstance(x, int)
+        assert x > acc
+        assert x < len(shape)
+
+        return x
+
+    reduce(lambda acc, x: _greater_than_reduce(acc, x), broadcast_dimensions, -1)
+
+    # shape must be broadcastable to
+    for idx, new_idx in enumerate(broadcast_dimensions):
+        assert a.shape[idx] == 1 or a.shape[idx] == shape[new_idx]
+
+def _broadcast_in_dim_aten(a, shape, broadcast_dimensions):
+    s = list(shape)
+    for broadcast_dimension in broadcast_dimensions:
+        s[broadcast_dimension] = -1
+
+    v = a
+    for idx, x in enumerate(s):
+        if x != - 1:
+            v.unsqueeze_(idx)
+
+    return v.expand(shape)
+
+
+def broadcast_in_dim(a, shape, broadcast_dimensions):
+    '''
+    Creates a view of t with the specified shape.
+
+    Allows adding dimensions of any length and broadcasting
+    dimensions of length one in t to any length.
+
+    The location of the broadcast dimensions must be specified
+    using the broadcast_dimensions argument. Changing the
+    relative order of dimensions is not supported.
+    '''
+
+    _broadcast_in_dim_checks(a, shape, broadcast_dimensions)
+    return _broadcast_in_dim_aten(a, shape, broadcast_dimensions)
+
+def broadcast(a, leading_lengths):
+  '''
+  A wrapper around broadcast_in_dim that just adds outermost dimensions
+  to a.
+  '''
+  shape = leading_lengths + a.shape
+  rank = len(leading_lengths) + len(a.shape)
+  return broadcast_in_dim(a, shape, tuple(range(len(leading_lengths), rank)))
+
+def _merge_dims_checks(a: Tensor, start: int, end: int):
+  assert isinstance(a, Tensor)
+
+  shape = a.shape
+  strides = a.stride()
+
+  _validate_idx(shape, start)
+  _validate_exclusive_idx(shape, end)
+
+  # Verifies end greater than start
+  assert end > start
+
+  for idx in range(start, end - 1):
+    assert strides[idx] == strides[idx + 1] * shape[idx + 1]
+
+def _merge_dims_aten(a: Tensor, start: int, end: int) -> Tensor:
+  # Short-circuits on null op
```

Comment:
Fair enough, but write a few of the preconditions of this function (`end > start`, at the very least).